File: examples/Reflection.py
#!/usr/bin/env python

# -------------
# Reflection.py
# -------------

print "Reflection.py"

class A :
    def f (self) :
        return "A.f()"

class B :
    def f (self) :
        return "B.f()"

def g (c) :
    x = c()
    return x.f()

assert g(A) == "A.f()"
assert g(B) == "B.f()"

d = globals()
assert type(d) is dict

a = d["A"]
assert a is A
assert g(a) == "A.f()"

b = d["B"]
assert b is B
assert g(b) == "B.f()"

print "Done."

File: examples/Reflection.java
// ---------------
// Reflection.java
// ---------------

interface I {
    String f ();}

final class A implements I {
    public String f () {
        return "A.f()";}}

final class B implements I {
    public String f () {
        return "B.f()";}}

final class Reflection {
    private static String g (Class c) {
        try {
            I x = (I) c.newInstance();
            return x.f();}
        catch (IllegalAccessException e) {
            e.printStackTrace();}
        catch (InstantiationException e) {
            e.printStackTrace();}
        return null;}

    public static void main (String[] args) {
        System.out.println("Reflection.java");

        {
        assert g(A.class) == "A.f()";
        assert g(B.class) == "B.f()";
        }

        {
        assert g(new A().getClass()) == "A.f()";
        assert g(new B().getClass()) == "B.f()";
        }

        try {
            assert g(Class.forName("A")) == "A.f()";
            assert g(Class.forName("B")) == "B.f()";}
        catch (ClassNotFoundException e) {
            e.printStackTrace();}

        System.out.println("Done.");}}
